File: app/src/main/java/cat/xojan/random1/ui/adapter/ProgramListAdapter.java
package cat.xojan.random1.ui.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.Collections;
import java.util.List;

import cat.xojan.random1.databinding.ProgramItemBinding;
import cat.xojan.random1.domain.entities.Program;
import cat.xojan.random1.domain.interactor.ProgramDataInteractor;
import cat.xojan.random1.viewmodel.ProgramViewModel;

public class ProgramListAdapter extends
        RecyclerView.Adapter<ProgramListAdapter.ProgramItemBindingHolder>  {

    private final ProgramDataInteractor mProgramDataInteractor;
    private List<Program> mProgramList;
    private Context mContext;

    public ProgramListAdapter(Context context, ProgramDataInteractor programDataInteractor) {
        mProgramList = Collections.emptyList();
        mContext = context;
        mProgramDataInteractor = programDataInteractor;
    }

    public void updateItems(List<Program> programs) {
        mProgramList = programs;
        notifyDataSetChanged();
    }

    @Override
    public ProgramItemBindingHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        ProgramItemBinding programItemBinding = ProgramItemBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false);
        return new ProgramItemBindingHolder(programItemBinding);
    }

    @Override
    public void onBindViewHolder(ProgramItemBindingHolder holder, int position) {
        Program program = mProgramList.get(position);
        holder.binding.setViewModel(
                new ProgramViewModel(mContext, program, mProgramDataInteractor));
        holder.binding.executePendingBindings();
    }

    @Override
    public int getItemCount() {
        return mProgramList.size();
    }

    static class ProgramItemBindingHolder extends RecyclerView.ViewHolder {

        private ProgramItemBinding binding;

        ProgramItemBindingHolder(ProgramItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/cat/xojan/random1/ui/adapter/ProgramListAdapter.kt
package cat.xojan.random1.ui.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import cat.xojan.random1.R
import cat.xojan.random1.domain.entities.Program
import cat.xojan.random1.domain.interactor.ProgramDataInteractor
import cat.xojan.random1.ui.activity.BrowseActivity
import com.squareup.picasso.Picasso
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.program_item.*
import java.util.*


class ProgramListAdapter(private val interactor: ProgramDataInteractor)
    : RecyclerView.Adapter<ProgramListAdapter.ViewHolder>() {

    var programs: List<Program> = emptyList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun getItemCount(): Int {
        return programs.size
    }

    override fun onBindViewHolder(holder: ViewHolder?, position: Int) {
        holder?.bind(programs[position])
    }

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): ViewHolder {
        val itemView = LayoutInflater.from(parent?.context)
                .inflate(R.layout.program_item, parent, false)
        return ViewHolder(itemView, interactor)
    }

    class ViewHolder(override val containerView: View,
                     val interactor: ProgramDataInteractor) : RecyclerView
    .ViewHolder(containerView), LayoutContainer {

        fun bind(item: Program) {
            itemView?.setOnClickListener {
                val isSection = interactor.isSectionSelected && item.sections.size > 1
                val intent = BrowseActivity.newIntent(itemView.context, item, isSection)
                itemView.context.startActivity(intent)
            }
            programTitle.text = item.title
            Picasso.with(itemView.context)
                    .load(item.imageUrl()+ "?w=" + getWeekOfTheYear())
                    .error(R.drawable.default_rac1)
                    .into(programImage)
        }

        private fun getWeekOfTheYear(): Int {
            val cal = Calendar.getInstance()
            return cal.get(Calendar.WEEK_OF_YEAR)
        }
    }
}